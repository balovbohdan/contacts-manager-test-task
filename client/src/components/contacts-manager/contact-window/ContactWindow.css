.main {}

.body {}

.phone {
    display: flex;
    align-items: center;
    justify-content: center;

    padding: 20px 10px;
    margin: 10px 0 15px;

    cursor: pointer;
    transition: var(--trans);
    border-radius: var(--border-radius);

    color: var(--color-yellow);
}

.phone:hover {
    background: rgba(255, 255, 255, .2);
}

.phone_txt {
    margin-left: 10px;
}

.phone_icon {
    width: 30px;
    height: 30px;

    background-size: contain;
    background-repeat: no-repeat;
    background-position: center center;
    background-image: url(~@img/logo.svg);
}

.tip {
    font-style: italic;
}

.actions {
    display: flex;
    align-items: center;
    justify-content: flex-start;
}

.actions_btn {
    width: 30px;
    height: 30px;

    display: flex;
    align-items: center;
    justify-content: center;

    cursor: pointer;
    margin-right: 10px;
    transition: var(--trans);
    border-radius: var(--border-radius);
    background-color: rgba(0, 0, 0, .3);
}

.actions_btn:hover {
    background-color: rgba(0, 0, 0, .4);
}

.actions_btnInnerWrapper {
    width: calc(100% - 15px);
    height: calc(100% - 15px);

    background-size: contain;
    background-repeat: no-repeat;
    background-position: center center;
}

.actions_btnInnerWrapper__edit {
    background-image: url(~@img/edit.svg);
}

.actions_btnInnerWrapper__remove {
    background-image: url(~@img/remove.svg);
}