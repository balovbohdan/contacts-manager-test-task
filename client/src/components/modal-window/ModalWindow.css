.main {
    z-index: 99999;
    position: absolute;

    width: 100vw;
    height: 100vh;

    top: 0;
    left: 0;
    right: 0;
    bottom: 0;

    display: flex;
    align-items: center;
    justify-content: center;
}

.body {
    max-width: 450px;
    width: calc(100vw - 20px);

    min-height: 200px;
    max-height: calc(100vh - 40px);

    padding: 10px;
    margin: 20px auto;

    position: relative;
    border-radius: var(--border-radius);
    background-color: var(--color-light-blue);
    box-shadow: 0 20px 35px rgba(0, 0, 0, .3), 0 15px 12px rgba(0, 0, 0, .2);

    color: var(--color-white);
}

.head {
    position: relative;
    padding-bottom: 10px;

    display: flex;
    align-items: center;
    justify-content: flex-start;
}

.closeBtn {
    position: absolute;

    top: -20px;
    right: -20px;

    width: 40px;
    height: 40px;

    display: flex;
    align-items: center;
    justify-content: center;

    cursor: pointer;
    border-radius: 50%;
    transition: var(--trans);
    background-color: var(--color-red);
    box-shadow: 0 3px 10px rgba(0, 0, 0, .3), 0 5px 10px rgba(0, 0, 0,.2);
}

.closeBtn:hover {
    transform: rotate(10deg) scale(.9);
}

.closeBtn_img {
    width: calc(100% - 20px);
    height: calc(100% - 20px);

    background-size: contain;
    background-repeat: no-repeat;
    background-position: center center;
    background-image: url(~@img/close.svg);
}