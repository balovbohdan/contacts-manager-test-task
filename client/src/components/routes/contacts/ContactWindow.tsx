import * as React from 'react';

import {findContact} from '@lib/entities/contacts';
import {T, utils} from '@lib/entities/calls-history';
import {Contacts} from '@lib/entities/contacts/contact/types';

import {T as Actions} from '@storehouse/actions/contacts';
import {ContactWindow as Base} from '@components/contacts-manager';

type Props = {
    close:()=>void;
    call:Actions.Call;
    remove:Actions.RemoveContact;
    toggleEditWindow:Actions.ToggleEditContactWindow;

    contacts:Contacts;
    active:boolean|number;
    callsHistory:T.CallsHistory;
    fetchCallsHistory:Actions.FetchCallsHistory;
};

export const ContactWindow = (props:Props) => {
    const active = +props.active || false;

    if (typeof active !== 'number')
        return null;

    const contact = findContact(active, props.contacts);

    if (!contact)
        return null;

    const history = utils.filterCallsHistory(
        contact.id,
        props.callsHistory
    );

    return <Base
        id={contact.id}
        call={props.call}
        close={props.close}
        remove={props.remove}
        callsHistory={history}
        contacts={props.contacts}
        toggleEditWindow={props.toggleEditWindow}
        fetchCallsHistory={props.fetchCallsHistory}/>;
};