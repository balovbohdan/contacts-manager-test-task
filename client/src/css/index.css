@import url(reset.css);
@import url(fonts.css);
@import url(titles.css);
@import url(constants.css);
@import url(main.css);