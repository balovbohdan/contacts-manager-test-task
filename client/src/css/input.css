.label {
    font-size: .9rem;
}

.input {
    width: 100%;
    padding: 5px 0;
    border-bottom: 2px solid var(--color-yellow);

    font-size: 1.5rem;
    color: var(--color-white);
}

@media (max-width: 600px) {
    .input {
        font-size: 1.2rem;
    }

    .label {
        font-size: .8rem;
    }
}

@media (max-width: 400px) {
    .input {
        font-size: 1.1rem;
    }

    .label {
        font-size: .7rem;
    }
}