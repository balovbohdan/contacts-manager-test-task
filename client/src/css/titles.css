h1, h2, h3, h4, h5, h6 {
    font-size: 100%;
    font-family: var(--font-main);
}

h1 {
    font-size: 40px;
}

h2 {
    font-size: 35px;
}

h3 {
    font-size: 30px;
}

h4 {
    font-size: 25px;
}

h5 {
    font-size: 20px;
}

h6 {
    font-size: 17px;
}

@media (max-width: 550px) {
    h1 {
        font-size: 35px;
    }

    h3 {
        font-size: 25px;
    }
}

@media (max-width: 400px) {
    h1 {
        font-size: 30px;
    }

    h3 {
        font-size: 20px;
    }
}