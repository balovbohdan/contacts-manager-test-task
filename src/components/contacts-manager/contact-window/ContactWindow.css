.main {

}

.body {

}

.phone {
    display: flex;
    align-items: center;
    justify-content: center;

    padding: 20px 10px;
    margin: 10px 0 15px;

    cursor: pointer;
    transition: var(--trans);
    border-radius: var(--border-radius);

    color: var(--color-yellow);
}

.phone:hover {
    background: rgba(255, 255, 255, .2);
}

.phone_txt {
    margin-left: 10px;
}

.phone_icon {
    width: 30px;
    height: 30px;

    background-size: contain;
    background-repeat: no-repeat;
    background-position: center center;
    background-image: url(~@img/logo.svg);
}

.tip {
    font-style: italic;
}