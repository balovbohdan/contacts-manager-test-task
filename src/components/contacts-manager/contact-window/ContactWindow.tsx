import * as React from 'react';
import classnames from 'classnames';

import {ModalWindow} from '@components/modal-window';

import * as T from './types';
import css from './ContactWindow.css';
import {CallsHistory} from './calls-history';

export const ContactWindow = (props:T.Props) => {
    const {name, img, phone, tip} = props.contacts[props.id];

    return (
        <ModalWindow close={props.close} title={name}>
            <Body
                img={img}
                tip={tip}
                phone={phone}
                id={props.id}
                call={props.call}
                remove={props.remove}
                contacts={props.contacts}
                callsHistory={props.callsHistory}
                toggleEditWindow={props.toggleEditWindow}
                fetchCallsHistory={props.fetchCallsHistory}/>
        </ModalWindow>
    );
};

const Body = (props:T.BodyProps) => {
    const {id, tip} = props;

    const doCall = () => props.call({
        contactId: id
    });

    return (
        <div className={css.body}>
            <Actions
                id={id}
                remove={props.remove}
                toggleEditWindow={props.toggleEditWindow}/>
            <div title='Call' className={css.phone} onClick={doCall}>
                <div className={css.phone_icon}/>
                <h5 className={css.phone_txt}>{props.phone}</h5>
            </div>
            {tip && <p className={css.tip}>{tip}</p>}
            <CallsHistory
                contactId={id}
                contacts={props.contacts}
                history={props.callsHistory}
                fetchCallsHistory={props.fetchCallsHistory}/>
        </div>
    );
};

const Actions = ({id, remove, toggleEditWindow}:T.ActionsProps) => {
    const doRemove = () => remove({ id });
    const openEditWindow = () => toggleEditWindow({ id });

    const editClassName = classnames(
        css.actions_btnInnerWrapper,
        css.actions_btnInnerWrapper__edit
    );

    const removeClassName = classnames(
        css.actions_btnInnerWrapper,
        css.actions_btnInnerWrapper__remove
    );

    return (
        <div className={css.actions}>
            <div onClick={openEditWindow} className={css.actions_btn}>
                <div className={editClassName}/>
            </div>
            <div onClick={doRemove} className={css.actions_btn}>
                <div className={removeClassName}/>
            </div>
        </div>
    );
};