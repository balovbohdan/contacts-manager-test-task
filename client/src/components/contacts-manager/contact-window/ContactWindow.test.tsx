import * as React from 'react';
import {mount, shallow} from 'enzyme';

import {ContactWindow} from './ContactWindow';
import * as contactWindowCss from './ContactWindow.css';
import * as modalWindowCss from '../../modal-window/ModalWindow.css';

describe('ContactWindow', () => {
    const propsDef = {
        id: 1,
        callsHistory: {},
        call: jest.fn(),
        close: jest.fn(),
        remove: jest.fn(),
        fetchCallsHistory: jest.fn(),
        toggleEditWindow: jest.fn(),
        contacts: [{
            id: 1,
            name: 'name',
            img: '/user.svg',
            phone: '+380959182534'
        }]
    };

    it('Should match snapshot', () => {
        const props = { ...propsDef };
        const window = shallow(<ContactWindow {...props}/>);

        expect(window).toMatchSnapshot();
    });

    it('Should rect on close btn click', () => {
        const props = {
            ...propsDef,
            close: jest.fn()
        };

        const window = mount(<ContactWindow {...props}/>);

        window.find(`.${modalWindowCss.closeBtn}`).simulate('click');
        expect(props.close).toBeCalledTimes(1);
    });

    it('Should close after empty space click', () => {
        const props = {
            ...propsDef,
            close: jest.fn()
        };

        const window = mount(<ContactWindow {...props}/>);

        window.find(`.${modalWindowCss.main}`).at(0).simulate('click');
        expect(props.close).toBeCalledTimes(1);
    });

    it('Should react on call btn click', () => {
        const props = {
            ...propsDef,
            call: jest.fn()
        };

        const window = mount(<ContactWindow {...props}/>);

        window.find(`.${contactWindowCss.phone_txt}`).simulate('click');
        expect(props.call).toBeCalledTimes(1);
    });

    it('Should react on remove btn click', () => {
        const props = {
            ...propsDef,
            remove: jest.fn()
        };

        const window = mount(<ContactWindow {...props}/>);

        window
            .find(`.${contactWindowCss.actions_btnInnerWrapper__remove}`)
            .simulate('click');

        expect(props.remove).toBeCalledTimes(1);
    });
});