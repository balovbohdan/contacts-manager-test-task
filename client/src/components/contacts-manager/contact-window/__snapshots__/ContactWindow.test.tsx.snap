// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactWindow Should match snapshot 1`] = `
<Component
  close={[MockFunction]}
  title="name"
>
  <Body
    call={[MockFunction]}
    callsHistory={Object {}}
    close={[MockFunction]}
    contacts={
      Array [
        Object {
          "id": 1,
          "img": "/user.svg",
          "name": "name",
          "phone": "+380959182534",
        },
      ]
    }
    fetchCallsHistory={[MockFunction]}
    id={1}
    img="/user.svg"
    phone="+380959182534"
    remove={[MockFunction]}
    toggleEditWindow={[MockFunction]}
  />
</Component>
`;
